internal/chain/bond_test.go

```diff
@@ -0,0 +1,290 @@
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := require.New(t)
+
+			b := NewBonder(memdb.New())
+			ts := tstate.New(0)
+			view := ts.NewView(state.CompletePermissions, newDB(t), 0)
+
+			if tt.maxBalance != nil {
+				r.NoError(b.SetMaxBalance(context.Background(), view, codec.EmptyAddress, *tt.maxBalance))
+			}
+
+			ok, err := b.Bond(context.Background(), view, tt.tx, tt.feeRate)
+			r.NoError(err)
+			r.Equal(tt.wantOk, ok)
+		})
+	}
+}
+
+func TestUnbond_NoBondCalled(t *testing.T) {
+	r := require.New(t)
+
+	b := NewBonder(memdb.New())
+
+	tx, err := chain.NewTxData(
+		&chain.Base{Timestamp: 123, ChainID: ids.Empty, MaxFee: 456},
+		nil,
+	).Sign(&NoAuthFactory{})
+	r.NoError(err)
+
+	r.NoError(b.Unbond(tx))
+}
+
+func TestUnbond_DuplicateUnbond(t *testing.T) {
+	r := require.New(t)
+
+	b := NewBonder(memdb.New())
+	ts := tstate.New(0)
+	view := ts.NewView(state.CompletePermissions, newDB(t), 0)
+
+	tx, err := chain.NewTxData(
+		&chain.Base{Timestamp: 123, ChainID: ids.Empty, MaxFee: 456},
+		nil,
+	).Sign(&NoAuthFactory{})
+	r.NoError(err)
+
+	ok, err := b.Bond(context.Background(), view, tx, 0)
+	r.True(ok)
+	r.NoError(err)
+
+	r.NoError(b.Unbond(tx))
+	r.NoError(b.Unbond(tx))
+}
+
+func TestUnbond_UnbondAfterFailedBond(t *testing.T) {
+	r := require.New(t)
+
+	b := NewBonder(memdb.New())
+	ts := tstate.New(0)
+	view := ts.NewView(state.CompletePermissions, newDB(t), 0)
+
+	tx, err := chain.NewTxData(
+		&chain.Base{Timestamp: 123, ChainID: ids.Empty, MaxFee: 456},
+		nil,
+	).Sign(&NoAuthFactory{})
+	r.NoError(err)
+
+	ok, err := b.Bond(context.Background(), view, tx, 1)
+	r.False(ok)
+	r.NoError(err)
+
+	r.NoError(b.Unbond(tx))
+}
+
+func TestSetMaxBalanceDuringBond(t *testing.T) {
+	r := require.New(t)
+	b := NewBonder(memdb.New())
+
+	ts := tstate.New(0)
+	view := ts.NewView(state.CompletePermissions, newDB(t), 0)
+	authFactory := &NoAuthFactory{}
+
+	tx1, err := chain.NewTxData(&chain.Base{Timestamp: 0}, nil).Sign(authFactory)
+	r.NoError(err)
+	tx2, err := chain.NewTxData(&chain.Base{Timestamp: 1}, nil).Sign(authFactory)
+	r.NoError(err)
+	tx3, err := chain.NewTxData(&chain.Base{Timestamp: 2}, nil).Sign(authFactory)
+	r.NoError(err)
+
+	r.NoError(b.SetMaxBalance(
+		context.Background(),
+		view,
+		codec.EmptyAddress,
+		uint64(tx1.Size()+tx2.Size()+tx3.Size()),
+	))
+
+	ok, err := b.Bond(context.Background(), view, tx1, 1)
+	r.NoError(err)
+	r.True(ok)
+
+	ok, err = b.Bond(context.Background(), view, tx2, 1)
+	r.NoError(err)
+	r.True(ok)
+
+	r.NoError(b.SetMaxBalance(context.Background(), view, codec.EmptyAddress, 0))
+
+	ok, err = b.Bond(context.Background(), view, tx3, 1)
+	r.NoError(err)
+	r.False(ok)
+
+	r.NoError(b.Unbond(tx1))
+	r.NoError(b.Unbond(tx2))
+}
+
+func newDB(t *testing.T) merkledb.MerkleDB {
+	db, err := merkledb.New(
+		context.Background(),
+		memdb.New(),
+		merkledb.Config{BranchFactor: 2},
+	)
+	require.NoError(t, err)
+	return db
+}
+
+type NoAuthFactory struct{}
```

Comment:
(optional for this PR / followup) Since we have an auth mock in the chain package, could we move this under `auth/authtest/` so that we can re-use instead of creating it in each place we need?